server/src/main/java/io/crate/replication/logical/action/TransportDropSubscriptionAction.java

```diff
@@ -90,4 +113,31 @@ protected ClusterState execute(ClusterState currentState, DropSubscriptionReques
     protected ClusterBlockException checkBlock(DropSubscriptionRequest request, ClusterState state) {
         return state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
     }
+
+    private Metadata.Builder removeSubscriptionSetting(Metadata.Builder mdBuilder, @Nonnull String subscriptionToDrop) {
+        Iterator<IndexService> indicesIterator = indicesService.iterator();
+        while (indicesIterator.hasNext()) {
+            IndexService indexService = indicesIterator.next();
+            var settings = indexService.getMetadata().getSettings();
+            var subscriptionName = REPLICATION_SUBSCRIPTION_NAME.get(settings); // Can be null.
+            if (subscriptionToDrop.equals(subscriptionName)) {
+
+                // Closes trackers of all shards of the index.
+                // Tracker closure removes retention lease.
+                shardReplicationService.stopTrackingIndexShards(indexService.shards());
```

Comment:
not synchronized on IndexService.shards since DROP SUBSCRIPTION is cluster state modifiers and there cannot be other operations running which could update shards(?)

Reply:
The shard tracker will run on any node containing a shard while this code here only runs on the master node.
Thus it will not stop any shard tracker.
We must listen for cluster state changes on every node instead and stop possible shard trackers on the node.
E.g. inside the `LogicalReplicationService.clusterChanges()`

Reply:
thx, will revisit this

Reply:
Done in https://github.com/crate/crate/pull/11937/commits/aff3726a5beebd5f8cf88314f61f3ebbd4d8eba6
initially I had a problem:
`LogicalReplicationService` now [needs](https://github.com/crate/crate/blob/4238e0ad39aea9081bd1e4b022e47b1b7221f57f/server/src/main/java/org/elasticsearch/node/Node.java#L540) `shardReplicationService` instance. 
I cannot create it like `LogicalRepService(..., ...,  new ShardReplicationService)` because of circular ref - ShardRepService itself needs LogicalRepService.

I applied similar approach with [restore](https://github.com/crate/crate/blob/4238e0ad39aea9081bd1e4b022e47b1b7221f57f/server/src/main/java/org/elasticsearch/node/Node.java#L589) & repositories services (inject not via constructor but set when we can) - and just got ShardReplicationService at the end directly from injector right after it's initialised for the sake of simplicity.